File: src/backend/AgentBackend.ts
import type { LoopEvent } from "src/types/events";
import type { LoopContext } from "src/loopContext";
import type { Provider } from "src/services/api/types";
import type { Message } from "src/types/message";
import { CurrentBackend } from "./CurrentBackend";

export interface BackendQueryParams {
  messages: Message[];
  ctx: LoopContext;
  provider: Provider;
  model?: string;
}

export interface AgentBackend {
  readonly name: "current" | "langgraph";
  query(params: BackendQueryParams): AsyncGenerator<LoopEvent, void>;
}

export function createBackend(): AgentBackend {
  const flag = process.env.DOCKER_AGENT_BACKEND ?? "current";
  if (flag === "langgraph") {
    // Lazy import to keep startup fast when defaulting to current.
    // Implementation added in Phase 2.
    const { LangGraphBackend } = require("./langgraph/LangGraphBackend") as {
      LangGraphBackend: new () => AgentBackend;
    };
    return new LangGraphBackend();
  }
  return new CurrentBackend();
}

File: python/src/docker_agent/backend/agent_backend.py
"""Agent backend abstraction + factory.

Parity: ``src/backend/AgentBackend.ts``.
"""

from __future__ import annotations

import os
from collections.abc import AsyncIterator
from typing import Any, Protocol, runtime_checkable

from pydantic import BaseModel, ConfigDict

from docker_agent.types.events import LoopEvent
from docker_agent.types.message import Message


class BackendQueryParams(BaseModel):
    model_config = ConfigDict(extra="forbid", arbitrary_types_allowed=True)
    messages: list[Message]
    ctx: Any  # LoopContext
    provider: Any  # Provider protocol — Any so test fakes validate
    model: str | None = None


@runtime_checkable
class AgentBackend(Protocol):
    name: str

    def query(self, params: BackendQueryParams) -> AsyncIterator[LoopEvent]: ...


def create_backend() -> AgentBackend:
    flag = os.environ.get("DOCKER_AGENT_BACKEND", "langgraph")
    if flag == "current":
        from docker_agent.backend.current_backend import CurrentBackend

        return CurrentBackend()
    from docker_agent.backend.langgraph.langgraph_backend import LangGraphBackend

    return LangGraphBackend()


__all__ = ["AgentBackend", "BackendQueryParams", "create_backend"]